splitgpg2/__init__.py

```diff
@@ -0,0 +1,804 @@
+        untrusted_args = untrusted_args.replace(b'+', b' ')
+        untrusted_args = re.sub(
+            rb'%[0-9A-F]{2}',
+            lambda m: bytes([int(m.group(0)[1:].decode('ascii'), 16)]),
+            untrusted_args,
+            flags=re.IGNORECASE
+        )
+        allowed_ascii = (string.printable + '\x0a').encode('ascii')
+        args = "Message from '{}':\n{}".format(
+            self.client_domain,
+            ''.join((chr(c) if c in allowed_ascii else '.')
+                    for c in untrusted_args)
+        )
+        return args.replace('%', '%25').\
+            replace('+', '%2B').\
+            replace('\n', '%0A').\
+            replace(' ', '+').\
+            encode('ascii')
+
+    async def command_RESET(self, untrusted_args: Optional[bytes]):
+        if untrusted_args:
+            raise Filtered
+        await self.agent_protocol.send_command(b'RESET', None)
+
+    async def command_OPTION(self, untrusted_args: Optional[bytes]):
+        if not untrusted_args:
+            raise Filtered
+
+        try:
+            untrusted_name, untrusted_value = untrusted_args.split(b'=', 1)
+        except ValueError:
+            untrusted_name = untrusted_args
+            untrusted_value = None
+        try:
+            action, opts = self.options[untrusted_name]
+            name = untrusted_name
+        except KeyError:
+            raise Filtered
+
+        if action == OptionType.override:
+            if opts:
+                option_arg = b'%s=%s' % (name, opts)
+            else:
+                option_arg = name
+        elif action == OptionType.verify:
+            if callable(opts):
+                verified = opts(untrusted_value=untrusted_value)
+            elif isinstance(opts, Pattern):
+                verified = (opts.fullmatch(untrusted_value) is not None)
+            else:
+                verified = (untrusted_value == opts)
+            if not verified:
+                raise Filtered
+            value = untrusted_value
+
+            if value:
+                option_arg = b'%s=%s' % (name, value)
+            else:
+                option_arg = name
+
+        elif action == OptionType.fake:
+            self.fake_respond(opts)
+            return
+
+        else:
+            raise Filtered
+
+        await self.agent_protocol.send_command(b'OPTION', option_arg)
+
+    async def command_AGENT_ID(self, untrusted_args: Optional[bytes]):
+        # pylint: disable=unused-argument
+        self.fake_respond(
+            b'ERR %d unknown IPC command' % GPGCode.UnknownIPCCommand)
+
+    async def command_HAVEKEY(self, untrusted_args: Optional[bytes]):
+        # upper keygrip limit is arbitary
+        args = self.verify_keygrip_arguments(1, 200, untrusted_args)
+        await self.agent_protocol.send_command(b'HAVEKEY', args)
+
+    async def command_KEYINFO(self, untrusted_args: Optional[bytes]):
+        # upper keygrip limit is arbitary
```

Comment:
copy paste error. Here it's not arbitary.

Reply:
Ok